### DISCUSSION TITLE
Can KafkaTransactionManager usable with a ReactiveTransactionManager ?

### DISCUSSION THREAD

**Discussion by xnart** [spring-projects/spring-kafka#2922] [Q&A] [answered]
I have a project which use R2dbc and spring-kafka altogether. I want to commit the Kafka transaction after R2dbc transaction. Normally it can be achieved as [documentation](https://docs.spring.io/spring-kafka/reference/tips.html#ex-jdbc-sync) suggests :
``` java
@Transactional("dstm")
public void someMethod(String in) {
    this.kafkaTemplate.send("topic2", in.toUpperCase());
    this.jdbcTemplate.execute("insert into mytable (data) values ('" + in + "')");
}
```

But when using R2dbcTransactionManager, KafkaTemplate throws `No transaction is in process; possible solutions: run the template operation within the scope of a template.executeInTransaction() operation, start a transaction with @Transactional before invoking the template method, run in a transaction started by a listener container when consuming a record`

I guess KafkaTransactionManager uses ThreadLocal while ReactiveTransactionManager uses Reactor Context. So they do not know each other's transactions. How can i adapt the KafkaTransactionManager to use Reactor Context ? Should i completely rewrite it ? 

Thanks in advance.

**Comment by artembilan** [comment, marked as answer]
Right, the `KafkaTemplate`  and its `KafkaProducer` are not Reactor-aware.
Therefore you cannot use them with `ReactiveTransactionManager`.

You may consider to use Rector Kafka instead: https://projectreactor.io/docs/kafka/release/reference/ `6.9. Transactional send`.
